### DISCUSSION TITLE
Custom Fonts

### DISCUSSION THREAD

**Discussion by Braehead** [HASwitchPlate/openHASP#317] [General]
I'm currently new to openHASP (but not this type of technology and programming), and using 0.6.2 and attempting to get custom fonts to work.  Not sure if I'm doing something wrong, or the methodology has changed since it was documented, or there is a bug located.

So I have followed the instructions on the Fonts section of the documentation.  Created a bin font file.  Not via the on-line method, as this appears to only produce the C code for compilation.  I can see no method to obtain a .bin version via on-line.  So used Linux tools (as described) to generate.

Uploaded the .bin file to the ESP via the web interface as display2412.bin.

Added the line to pages.jsonl, and uploaded:

{"page":1,"id":100,"obj":"label","x":0,"y":40,"h":40,"w":60,"text":"00.0","align":2,"text_color":"black","text_font":"display2412,"enabled":"false"}

But the display of 00.0 (after a restart) is in a basic and very small font.  Not what was uploaded and specified.

Any ideas? Or further information needed?

Many thanks.

**Comment by fvanroie** [comment]
Which ESP32 board are you using? External fonts require RAM to pre-load them, so PSram is almost a must except for tiny font files.
Do you have the link to the documentation page and online converter you used? I'd like to double-check the info and results.

Custom Fonts are in the process of being reworked #249. The methodology you are describing is correct, but is only available in 0.6.3-dev.
In v0.6.2 this method just won't work...

**Comment by Braehead** [comment]
Used the offline version detailed here: https://github.com/lvgl/lv_font_conv and installed with:

npm i lv_font_conv -g

Then used to convert:

lv_font_conv --bpp 2 --size 12 -o display2412.bin --format bin --font Display24.ttf --symbols C0123456789,.

The resultant file is 292 bytes.  Shouldn't be too large, even for PSRAM?  Actually I want 72px short font.  The source font was called open_24_display_st.ttf - just renamed above with a shorter name to convert.

Only need those digits,  So I'm not expecting this to be a serious stress.

Using Adafruit Hazzah Featherwing (ESP32) with their 2.4" TFT.  I'll see if I can get the 0.6.3 dev branch, is it available on GH?

When I used the LVFL online font converter (https://lvgl.io/tools/fontconverter), I could not get a binary output, only a C source.  Is there a different link?

**Comment by fvanroie** [comment]
292 bytes is not a problem, even for the Hazzah32 without PSram. I estimate you can add several kilobytes of fonts without any issue.

The official lvgl online font converter exports C code only. That is why we have provided our own [customized online binary font converter](https://openhasp.haswitchplate.com/0.6.3/design/fonts/#online-font-converter).
It will produce the .bin files as needed, but the offline font converter can yield the same result.

The [latest build ](https://github.com/HASwitchPlate/openHASP/actions/runs/2086040107)can always be found under the Actions tab on Github. Click the `featherwing` artifact near the bottom of the page.

**Reply by Braehead** [reply]
Thanks, I'll download an updated version this evening.

I have to be honest, I did not realise that the online font converter on that page of the documentation was actually usable, until I have now just clicked on it.  To me, it looked like a screen shot of the LVGL font conversion, and not an active page element.  I just scrolled through reading the documentation carefully - Doh.  I was looking for a dedicated link to a page with just that function.  Unless the mouse is over certain parts, it just looks like a graphic.  As I use a mouse and scroll wheel a lot, the pointer would have been towards the right and never changed.  Sorry about that.

**Reply by fvanroie** [reply]
No worries, I can see where you're coming from. That page includes a _patched_ version of the LVGL online font converter and Javascript/CSS is not my area of expertise. I thought it would be _*good enough*_ but clearly there is room for improvement...

It would be easier if the online LVGL converter had an option to export .BIN files too.

**Comment by Braehead** [comment]
Yes 0.6.3 dev appears to work.  I'll do a few more tests for you.  While I'm in there, would you like anything tested in this branch? and I'll see what I can do.

**Reply by fvanroie** [reply]
I'm not actively testing against the Featherwing hardware myself at the moment, so the confirmation it seems to work is already a good thing. Feel free to post any issues you may find.

**Comment by nagyrobi** [comment]
Try using TTF font files directly.

For example:
```
{"page":1,"id":2,"obj":"btn","x":20,"y":43,"w":200,"h":145,"toggle":true,"text":"\ue9c4","text_font":"hass-hue-icons98","mode":"break","align":1,"radius":20,"groupid":1,"text_color":"white","text_color2":"white","text_color1":"yellow","text_color3":"yellow"}
```
Upload to flash the `hass-hue-icons.ttf` from [here](https://github.com/arallsopp/hass-hue-icons/tree/main/font).

Should give you something like:
![kép](https://user-images.githubusercontent.com/1550668/161927701-f5502c00-a869-47c4-985e-96de598e7cad.png)

Note that in `"text_font":"hass-hue-icons98"` the `hass-hue-icons` is the name of the TTF font file without extension, and `98` is the font size.

**Reply by fvanroie** [reply]
I have fixed some TTF font bugs in the latest builds. Freetype is enabled again and can be used as you described.
It should make the usage of fonts a lot easier...

**Reply by Braehead** [reply]
Thanks.  I'll try in the next few days.  I did a basic touch control panel and have been doing all the Node Red background functions and logic.  Now ready to upgrade and prettyfy the OpenHASP part.  I'll write up the project over the next few weeks and make it available.

Basically a truly programmable/configurable room temperature controller and monitor.  Not a thermostat or relay (this is handled elsewhere in the the Home Automation setup).  But a temperature control and information touch panel for Home Automation.  I believe this to be the only such unit for Home Automation purposes that isn't a bodge / hack of a commercial unit.  But specifically chosen and implemented.  A mixture of an openHASP panel, MQTT, Node Red and Tasmota.  So should be compatible with many Home Automation systems to some degree,

**Reply by fvanroie** [reply]
Today some more TTF bugs were fixed an pushed to the 0.6.3-dev build, this should be pretty much complete now.

**Reply by nagyrobi** [reply]
Qualifies as a full replacement of the built-in font!

**Comment by zagi988** [comment]
How do you know what code to use for the icon?
Where do you find the codes?
I'm trying to use mdicons.ttf and am using the cheatsheet with codes but when using it like this:
{"page":1,"id":4,"obj":"btn","x":15,"y":145,"w":90,"h":90,"toggle":true,"text_font":"mdicons32","align":"center","radius":10,"value_str":"Kuhinja L1","value_align":"top","value_ofs_y":-30}

and
- obj: "p1b4"
      properties:
        "val": '{{ 1 if is_state("light.kuhinja_l1", "on") else 0 }}'
        "text": '{{ "\uF091D" if is_state("light.kuhinja_l1", "on") else "\uF091D" | e }}'
        
It doesn't display the icon. What am I doing wrong?

**Reply by zagi988** [reply]
Nvrm, got it. Missed the part about surrogate pair.
